relation/src/proof_linking/arithmetization.rs

```diff
@@ -59,7 +59,11 @@ impl GroupLayout {
 impl<F: PrimeField> PlonkCircuit<F> {
     /// Generate a layout of the circuit, including where proof-linking gates
     /// will be placed
-    pub fn gen_circuit_layout(&self) -> Result<CircuitLayout, CircuitError> {
+    pub fn gen_circuit_layout(&mut self) -> Result<CircuitLayout, CircuitError> {
+        if let Some(layout) = &self.layout {
+            return Ok(layout.clone());
+        }
+
```

Comment:
IMO, when we later reuse the circuit layout for e.g. generating linking proofs, i would opt not to use the `gen_circuit_layout` method there, and instead directly get the layout via `self.layout.expect(...)`. This feels clearer to me because we expect the layout to be generated by that point, and it is more accurate to say we're referencing it than generating it.

this would mean removing this conditional here, or better yet:
```suggestion
    pub fn gen_circuit_layout(&mut self) -> Result<CircuitLayout, CircuitError> {
        if &self.layout.is_some() {
            return CircuitError::LayoutAlreadyGenerated;
        }

```

this would ensure that we are not generating the layout more than once

Reply:
Hm, this method should definitely be safe so we need this check. Don't agree with returning an error here, no error has occurred! And while callers can certainly access the layout directly, it's odd to have an implicit semantic that a method *may only ever be called once* unless we absolutely need it -- in this case we don't. For this reason I feel the method should be safe to call even if callers may have a cleaner option